### PR TITLE
Fix invalid recursion errors for structs

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/3275
Fixes https://github.com/ethereum/solidity/issues/3323

### CODE REVIEW

File: libsolidity/analysis/PostTypeChecker.cpp [hunk 1]
```diff
@@ -87,20 +89,24 @@ bool PostTypeChecker::visit(Identifier const& _identifier)
...
-	}
-	return nullptr;
+		// Iterating through the dependencies needs to be deterministic and thus cannot
+		// depend on the memory layout.
+		// Becauso of that, we sort by AST node id.
+		std::vector<VariableDeclaration const*> dependencies(
+			m_constVariableDependencies[&_variable].begin(),
+			m_constVariableDependencies[&_variable].end()
+		);
+		std::sort(dependencies.begin(), dependencies.end(), [](VariableDeclaration const* _a, VariableDeclaration const* _b) -> bool
```
Comment (axic) [inline libsolidity/analysis/PostTypeChecker.cpp]: No need for explicit namespace here.

File: libsolidity/analysis/PostTypeChecker.cpp [hunk 2]
```diff
@@ -87,20 +89,24 @@ bool PostTypeChecker::visit(Identifier const& _identifier)
...
-		set<VariableDeclaration const*> seen(_seen);
-		seen.insert(_startingFrom);
-		for (auto v: m_constVariableDependencies[_startingFrom])
-			if (findCycle(v, seen))
-				return v;
-	}
-	return nullptr;
+		// Iterating through the dependencies needs to be deterministic and thus cannot
+		// depend on the memory layout.
+		// Becauso of that, we sort by AST node id.
```
Comment (axic) [inline libsolidity/analysis/PostTypeChecker.cpp]: Typo.

File: libdevcore/Algorithms.h [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+		if (m_processed.count(&_vertex))
+			return nullptr;
+		else if (m_processing.count(&_vertex))
+			return m_firstCycleVertex = &_vertex;
+		m_processing.insert(&_vertex);
+
+		m_depth++;
+		m_visit(_vertex, *this);
+		m_depth--;
+		if (m_firstCycleVertex && m_depth == 1)
```
Comment (axic) [inline libdevcore/Algorithms.h:60]: What case is this workaround for?

**Review by axic — CHANGES_REQUESTED** [review]
Please rebase.

**Comment by axic** [comment]
Perhaps this will also fix #3322 and #3323 ?

File: libdevcore/Algorithms.h (hunk 3 above)
Comment (chriseth) [reply libdevcore/Algorithms.h:60]: It is not a workaround, this is to properly report where the recursion starts. If you have `a -> b -> c -> d -> c`, then it should report that there is a cycle that starts at `c`. I think...

This was basically needed to keep the error messages as they were.

Comment (chriseth) [reply libdevcore/Algorithms.h:60]: Ah no, it should say it is a cycle that goes via `d`.

Comment (chriseth) [reply libdevcore/Algorithms.h:60]: Via `b`!

**Comment by chriseth** [comment]
Rebased.

**Comment by chriseth** [comment]
Will check if it fixes the two mentioned issues.

**Comment by chriseth** [comment]
It fixes the second one. For the first one (https://github.com/ethereum/solidity/issues/3322), the RAM blowup is gone, but it still takes very long to compile. Not sure if we should consider this fixed.

**Review by axic — APPROVED** [review]

**Pull request merged**